### PR TITLE
[no-implicit-copy] Update SILGen/move checker to work with new patterns from copyable_to_moveonly and friends.

### PR DESCRIPTION
**Pull request by gottesmm**
This involved doing the following:

1. Update the move only checker to look for new patterns.

2. Teach emitSemanticStore to use a moveonlywrapper_to_copyable to store moveonly
   values into memory. The various checkers will validate that this code is
   correct.

3. When emitting an apply, always unwrap move only variables. In the
   future, I am going to avoid this if a parameter is explicitly marked as also
   being moveonly (e.x.: @moveOnly parameter or @noEscape argument).

4. Convert from moveOnly -> copyable on return inst automatically in SILGen.

5. Fix SILGenLValue emission so we emit an error diagnostic later rather than
   crash. This is needed to keep SILGen emitting move only addresses (that is no
   implicit copy address only lets) in a form that the move only checker then
   will error upon. Without this change, SILGen crashes instead of emitting an
   error diagnostic in the following test:
   .//test/SILOptimizer/move_only_checker_addressonly_fail.swift

Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves SR-NNNN.

### CODE REVIEW

**Comment by gottesmm** [comment]
@jckarter I forgot to upload the trivial type eliminator. I added that in the 2nd commit I added here and also added a small fix that I discovered when writing some Interpreter tests for this.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

File: test/SILOptimizer/move_only_checker_trivial.swift [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+        } else {
+            let y = x2 // expected-note {{consuming use}}
+            let _ = y
+        }
+    }
+}
+
+public func aggGenericStructAccessField<T>(_ x: AggGenericStruct<T>) {
+    @_noImplicitCopy let x2 = x // expected-error {{'x2' consumed more than once}}
+    print(x2.lhs) // expected-note {{consuming use}}
```
Comment (jckarter) [inline test/SILOptimizer/move_only_checker_trivial.swift:387]: Projecting a stored property, or more generally, calling a `get` or `read` accessor with a guaranteed base, should not be a consuming use.

Comment (gottesmm) [reply test/SILOptimizer/move_only_checker_trivial.swift:387]: The reason this is a consuming use is b/c of print. Print consumes values unfortunately due to the way we construct the array.

File: test/SILOptimizer/move_only_checker_trivial.swift [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        print(x2.lhs) // expected-note {{consuming use}}
+    }
+}
+
+/////////////////////////////////
+// No Implicit Copy Attributes //
+/////////////////////////////////
+
+public func klassNoImplicitCopyArgument(@_noImplicitCopy _ x: Trivial) -> Trivial {
+    return x
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply test/SILOptimizer/move_only_checker_trivial.swift:398]: I actually have @_noImplicitCopy arguments being +1 currently. I am looking at changing that (I can do it in this PR)

**Comment by gottesmm** [comment]
@jckarter Hows this sound: can we get in this PR and then in follow on PRs I will do the following:

1. Remove the typelowering bit for move only (leaving in typelowering itself though)
2. I am going to rename MoveOnlyType to MoveOnlyWrappedType
3. I am going to change the owned no implicit copy ABI change to not be there and add a test that shows the behavior is as expected for non-trivial types as well.
4. I am going to make it so that no-implicit-copy does not apply to fields of the underlying struct since it is a getter and the type will be copyable.

**Comment by jckarter** [comment]
That works for me.

**Review by jckarter — APPROVED** [review]

**Pull request merged**